Display keybindings in UI consistently and add options to allow players to customize how the keybindings are displayed

strategineer opened the pull request:
<!-- HOW TO USE: Under each "#### Heading" below, enter information relevant to your pull request.
Leave the headings unless they don't apply to your PR.

Please read carefully and don't delete the comments delimited by "< !--" and "-- >"
Once a pull request is submitted, automatic stylistic and consistency checks will be performed on the PR's changes.
The results of these can be either seen under the "Files changed" section of a PR or in the check's details.

NOTE: Please grant permission for repository maintainers to edit your PR.  It is EXTREMELY common for PRs to be held up due to trivial changes being requested and the author being unavailable to make them. -->

#### Summary
Interface "Display keybindings in UI consistently and add options to allow players to customize how the keybindings are displayed"
<!-- This section should consist of exactly one line, edit the one above.
1. Replace the word "Category" with one of these words: Features, Content, Interface, Mods, Balance, Bugfixes, Performance, Infrastructure, Build, I18N.
2. Replace the text inside the quotes with a brief description of your changes.
Or if you don't want a changelog entry, replace the whole line with just the word "None" (with no quotes).
For more on the meaning of each category, see:
https://github.com/CleverRaven/Cataclysm-DDA/blob/master/doc/CHANGELOG_GUIDELINES.md
If approved and merged, your summary will be added to the project changelog:
https://github.com/CleverRaven/Cataclysm-DDA/blob/master/data/changelog.txt -->

#### Purpose of change

F*xes most of #58846 but not all...

This PR will:
- Make the code cleaner.
- Improve UX by making it (more) clear which keybindings the player can press at any time in a way that they're more likely to recognize at a glance.
- Ensure that we never display hardcoded keybindings to the player to prevent the situation where the player would rebind some keys but they're hardcoded in the UI so we'll end up displaying wrong keybindings (bad UX).
- Allow players to customize how keybindings are displayed throughout the UI to suit their needs (eg. "[x] Examine", "x: Examine", "(x) Examine", "x - Examine", "E[x]amine").
- Make extensive use of `fold_and_print` for displaying keybindings to allow them to show up well (in theory) no matter what language they're displayed in.

<!-- With a few sentences, describe your reasons for making this change.  If it relates to an existing issue, you can link it with a # followed by the GitHub issue number, like #1234.  If your pull request *fully* resolves an issue, include the word "Fix" or "Fixes" before the issue number, like: Fixes #xxxx
If there is no related issue, explain here what issue, feature, or other concern you are addressing.  If this is a bugfix, include steps to reproduce the original bug, so your fix can be verified. -->

#### Describe the solution

<!-- How does the feature work, or how does this fix a bug?  The easier you make your solution to understand, the faster it can get merged. -->

- Create some `input_context::get_hint*` methods built on the existing `get_desc` methods that provide a more complete interface for generating the strings to display.
- Replace all existing uses of `get_desc` outside of the input.cpp with calls to the more complete `get_hint` methods (greatly simplifying UI code in the process).
- Find hardcoded keybindings and replace the vast majority of them with calls to `get_hint*` methods (passing any relevant input_context object around as needed) so that the keybindings are displayed consistently and always display an actual keybinding that the player has set.
- Insert extra bits of code into the existing `shortcut_{print|text}` and `press_x` methods that are harder to refactor to make the keybindings generated by them appear consistently with the keybindings generated through `get_hint*` methods  (This refactoring can be tackled in a subsequent PR to make this one more reasonable to review).
- Add `input_context::get_hint_color*` methods (NB: these should probably be in the `color.h` file) that take a state as a parameter (like ENABLED, DISABLED, TOGGLED_ON, TOGGLED_OFF, HIGHLIGHTED) and try to use them consistently throughout the code as needed such that we have one single place where we can swap all keybindings-related colors around.
- Fix minor UI issues/inconsistencies while I was in the neighborhood.

#### Describe alternatives you've considered

<!-- Explain any alternative solutions, different approaches, or possibilities you've considered using to solve the same problem. -->

None.

#### Testing

<!-- Describe what steps you took to test that this PR resolved the bug or added the feature, and what tests you performed to 
make sure it didn't cause any regressions.  Also include testing suggestions for reviewers and maintainers. -->

- Each changed menu (except the basecamp UI code because I don't know how to trigger it :() was tested at least once after making changes to ensure that nothing was broken.
- WIP I'm doing a second pass now of all the changed screens (most of them) to make sure nothing is broken but any help with testing this would be appreciated.

#### Additional context

To Review:
The majority of the important changes can be found in `input.*`, `options.cpp` and `output.*`. All the other changes stem from there and, for the most part, are relatively simple. I've attached screenshots for each changed UI screen, you can see them by scrolling down!

Here are some screenshots for different keybinding hint settings:

Default settings:
![WindowsTerminal_I837azrnMt](https://user-images.githubusercontent.com/2420411/228027047-f78bd8fb-20d6-4254-8355-81ab8fe66f43.png)
![WindowsTerminal_Wb4qL7nOkJ](https://user-images.githubusercontent.com/2420411/228027069-0ef0da71-32de-4796-985d-e9b077c0d376.png)

settings example \#1:
![WindowsTerminal_y03FKJNmFe](https://user-images.githubusercontent.com/2420411/228027072-e06ef41b-f4d4-4610-9aea-13a7bd49de8c.png)
![WindowsTerminal_w4UwxOrAWj](https://user-images.githubusercontent.com/2420411/228027085-1af64176-6c98-4b30-8013-f5935f8cbb98.png)

settings example \#2:
![WindowsTerminal_4ZpIxUVmye](https://user-images.githubusercontent.com/2420411/228027094-a15bdebd-71ae-4b2f-b503-43516775be83.png)
![WindowsTerminal_PxqIkHzFyK](https://user-images.githubusercontent.com/2420411/228027101-cfa70633-50a1-45d3-8a07-ffae224785f4.png)

settings example \#3:
![WindowsTerminal_TwlGKwnL2A](https://user-images.githubusercontent.com/2420411/228027110-a7a572ab-a05c-4fa0-9201-ed862065b09e.png)
![WindowsTerminal_Udqlq5im11](https://user-images.githubusercontent.com/2420411/228027121-74951037-2902-441c-adcd-aa33cbda820a.png)

strategineer committed:
Display keybindings in UI consistently and add options to allow players to customize
how the keybindings are displayed

Qrox commented on src/output.cpp:
```diff
@@ -2699,9 +2717,22 @@ std::pair<std::string, nc_color> rad_badge_color( const int rad )
-std::string formatted_hotkey( const std::string &hotkey, const nc_color text_color )
+std::string keybind_hint( const std::string &hotkey )
+{
+    // todo Add option to select between [x], x: and e[x]amine?
```
FYI [this method of `input_context`](https://github.com/CleverRaven/Cataclysm-DDA/blob/dc9b4941013aafa6b8cae9dff13aa3695c5a75f2/src/input.cpp#L1115-L1120) have something similar to what you want here, but for registered input actions.

Qrox commented on src/inventory_ui.cpp:
```diff
@@ -299,12 +299,12 @@ std::string enum_to_string<inventory_selector::uimode>( inventory_selector::uimo
     switch( mode ) {
         case inventory_selector::uimode::hierarchy:
-            return "hierarchy";
+            return "Hierarchy";
```
This is used for serialization of the enum, and is not marked for translation. You'd better create a separate function for the translated strings.

strategineer commented on src/worldfactory.cpp:
```diff
@@ -1927,12 +1908,6 @@ int worldfactory::show_worldgen_basic( WORLD *world )
                 int lvl = wg_slevels[sel_opt - 1] + ( action == "LEFT" ? -1 : 1 );
                 wg_slevels[sel_opt - 1] = clamp<int>( lvl, 0, wg_sliders[sel_opt - 1]->count() - 1 );
                 wg_sliders[sel_opt - 1]->apply_opts( wg_slevels[sel_opt - 1], world->WORLD_OPTIONS );
-            } else if( sel_opt > static_cast<int>( wg_sliders.size() ) ) {
```
worldgen option "buttons" are displayed vertically now, so don't need this anymore

Qrox commented on src/inventory_ui.cpp:
```diff
@@ -560,11 +560,8 @@ nc_color inventory_entry::get_invlet_color() const
-        return c_yellow;
-    } else {
-        return c_white;
     }
+    return c_yellow;
```
I think yellow and white are used to distinguish user-assigned and auto-assigned inventory letters.

strategineer replied:
I wasn't sure about this change because I don't play with the auto-assigned inventory letters option (sidenote: it assigns the keys I use for navigation and it assigns digits preventing me from dropping/inserting X items at a time making the feature less than useful. Although both of these problems are bugs that can be fixed separately IMO).

Is it useful to the player for us to distinguish between user-assigned and auto-assigned inventory letters/keybindings?

NB: Thanks for the help @Qrox, it's much appreciated!

Qrox replied:
I haven't used it in a while but the last time I used it user-assigned ones behave differently. I don't remember the details but I think user-assigned ones persist even if you drop the item and pick it back up, while auto-assigned ones can change after dropping and picking up.

strategineer replied:
ahhhhhh might as well keep the color scheme as-is then, I'm already making enough changes here :)

Qrox commented on src/gamemode_defense.cpp:
```diff
@@ -1228,9 +1232,10 @@ void draw_caravan_borders( const catacurses::window &w, int current_window )
 
-    // Quick reminded about help.
+    // Quick reminder about help.
     // NOLINTNEXTLINE(cata-text-style): literal question mark
```
```suggestion
```

strategineer commented on src/action.cpp:
```diff
@@ -536,16 +536,6 @@ std::string press_x( action_id act, const std::string &key_bound_pre,
     input_context ctxt = get_default_mode_input_context();
     return ctxt.press_x( action_ident( act ), key_bound_pre, key_bound_suf, key_unbound );
 }
-std::optional<std::string> press_x_if_bound( action_id act )
```
function used once in src/help.cpp, deleting it. It'll be easier to refactor the rest of these `press_x` functions later on.

Qrox commented on src/input.cpp:
```diff
@@ -1145,38 +1157,205 @@ std::string input_context::get_desc(
+    if( state != keybinding_hint_state::NONE ) {
+        std::string c_sep = get_tag_from_color( get_hint_color_for_separator( state ) );
+        std::string c = get_tag_from_color( get_hint_color( state ) );
+        inline_fmt = to_translation( "keybinding",
+                                     string_format( "%s%s</color>%s%s</color>%s%s</color>%s%s</color>%s%s</color>", c, "%1$s", c_sep,
+                                             lp_sep, c_key, "%2$s", c_sep, rp_sep, c, "%3$s" ) );
+        separate_fmt = to_translation( "keybinding",
+                                       string_format( "%s%s</color>%s%s</color>%s%s</color>%s%s</color>", c_sep, lb_sep, c_key, "%1$s",
+                                               c_sep, rb_sep, c, "%2$s" ) );
```
The arguments to `to_translation` have to be string literals or it won't work (the string extractor won't know what runtime values the format arguments will take).

Also formatting a string with format specifiers as arguments to form a new format string feels too hacky to me, and it will likely break if some of the arguments happen to contain a sequence that can be interpreted as a format specifier.

I suggest passing fixed format strings and the needed variables as arguments to the final `get_desc` call instead and format the text in one go there.

Might want to use `colorize` instead of color tag string too to avoid confusing the translators with unpaired color tags.

strategineer replied:
Weird, I responded to you this morning but I guess I never hit return? AHHHHHHH

This code was absolutely atrocious :) I refactored it in the way you suggested, thanks!

Qrox commented on src/input.cpp:
```diff
@@ -1145,38 +1157,205 @@ std::string input_context::get_desc(
+                                       string_format( "%s%s</color>%s%s</color>%s%s</color>%s%s</color>", c_sep, lb_sep, c_key, "%1$s",
+                                               c_sep, rb_sep, c, "%2$s" ) );
+    } else {
+        //  %1$s: action description text before key,
+        //~ %2$s: key description,
+        //~ %3$s: action description text after key.
+        inline_fmt = to_translation( "keybinding", string_format( "%s%s%s%s</color>%s%s", "%1$s", lp_sep,
+                                     c_key,
+                                     "%2$s", rp_sep, "%3$s" ) );
+        // \u00A0 is the non-breaking space
+        //~ %1$s: key description,
+        //~ %2$s: action description.
+        separate_fmt = to_translation( "keybinding", string_format( "%s%s%s</color>%s%s", lb_sep, c_key,
+                                       "%1$s",
+                                       rb_sep, "%2$s" ) );
```
Same here. Also the comments starting with "//~" are intended for translators and only apply to the extracted format strings, so the translators will not know what "%1$s" etc refer to after this change.

Qrox commented on src/input.cpp:
```diff
@@ -1145,38 +1157,205 @@ std::string input_context::get_desc(
 
+std::string replace_spaces_with_non_breaking( std::string s )
+{
+    return std::regex_replace( s, std::regex( " " ), "\u00A0" );
```
Use regex for a simple substring replacement might be overkill.

strategineer replied:
The problem I was running into is that `std::replace` doesn't handle strings, it only works with characters which is a problem when dealing with unicode values.

Should I replace this call to `regex_replace` with custom string manipulation code?

Qrox replied:
Huh, didn't know that the standard library does not provide string replacement. Yeah I think a custom string manipulation function would be better here.

strategineer replied:
Done!

strategineer committed:
Don't register "toggle_pixel_minimap" keybind when TILES=0

... because it doesn't do anything and it shouldn't show up
as a keybind hint in the "look around" menu.

strategineer commented on src/wish.cpp:
![WindowsTerminal_6UAplGZAlc](https://user-images.githubusercontent.com/2420411/228043337-eb365e91-89fa-45fc-a968-265ed3b91cfd.png)
![WindowsTerminal_7SSSgAXStn](https://user-images.githubusercontent.com/2420411/228043351-e2618cc9-c0bc-4278-a0d6-deab428bbccb.png)

strategineer commented on src/worldfactory.cpp:
![image](https://user-images.githubusercontent.com/2420411/228043475-c32de557-bd0c-4d93-806d-9d4d049039e9.png)

strategineer commented on src/veh_interact.cpp:
![image](https://user-images.githubusercontent.com/2420411/228043924-09a5467f-701e-4b75-b463-a0889bf087a4.png)

strategineer committed:
Cleanup, fix main menu width when not inlining keybinds and make keybinds display consistently in help menu

strategineer commented on src/armor_layers.cpp:
![image](https://user-images.githubusercontent.com/2420411/228067692-b1ace9cd-2486-4341-8848-8bd22af0a72f.png)
![image](https://user-images.githubusercontent.com/2420411/228067729-ac515870-209a-4e1a-8147-74d0ac228f39.png)

strategineer commented on src/advanced_inv.cpp:
![image](https://user-images.githubusercontent.com/2420411/228067849-22e1a3d9-9c6e-48cf-ba27-a6aa10684bd3.png)

![image](https://user-images.githubusercontent.com/2420411/228067969-f0425b21-82e5-4d47-b840-8615731aef4d.png)

strategineer commented on src/auto_note.cpp:
![image](https://user-images.githubusercontent.com/2420411/228068050-9c24e027-1b9f-412d-84c1-763fcb7548b2.png)

strategineer commented on src/auto_pickup.cpp:
![image](https://user-images.githubusercontent.com/2420411/228068099-d69b62b1-c0b1-4dfa-aade-8092d86b4834.png)

strategineer commented on src/avatar.cpp:
![image](https://user-images.githubusercontent.com/2420411/228068188-1bfe72a5-c966-4f8b-94d2-924561e4e562.png)

strategineer commented on src/bionics_ui.cpp:
![image](https://user-images.githubusercontent.com/2420411/228068622-a4a342b5-1c35-4491-86d9-a4c5197235d9.png)

strategineer commented on src/color.cpp:
![image](https://user-images.githubusercontent.com/2420411/228068745-2973a600-0187-4a97-9b3b-72ac4e43fca3.png)

strategineer commented on src/construction.cpp:
![image](https://user-images.githubusercontent.com/2420411/228069206-9c11d80b-d79c-4e5f-a3b7-046613e7a54e.png)

strategineer commented on src/crafting_gui.cpp:
![image](https://user-images.githubusercontent.com/2420411/228069970-0a3434e7-210d-4137-9fd7-571342b9ec67.png)

strategineer commented on src/debug_menu.cpp:
![image](https://user-images.githubusercontent.com/2420411/228070052-5cef3a29-73aa-4b04-8fe2-b8d1f8403a0c.png)

strategineer committed:
Normalize dialogue UI

strategineer commented on src/dialogue_win.cpp:
![image](https://user-images.githubusercontent.com/2420411/228077181-e6a3f768-88ad-4c63-aa38-f49923a496ab.png)

strategineer commented on src/diary_ui.cpp:
![image](https://user-images.githubusercontent.com/2420411/228077241-cc127c9f-c45d-49c0-8a65-f29cf1d5b584.png)

strategineer commented on src/editmap.cpp:
![image](https://user-images.githubusercontent.com/2420411/228078564-48ccd8db-596e-4364-b9ba-beca6f42052d.png)

strategineer commented on src/game.cpp:
![image](https://user-images.githubusercontent.com/2420411/228078687-7eec2d5f-c779-4d02-a858-93c57bd0aede.png)
![image](https://user-images.githubusercontent.com/2420411/228078741-f7221020-c41b-4bb0-8c0b-a2de2ecf29d0.png)
![image](https://user-images.githubusercontent.com/2420411/228078859-6b8ee3fd-a8af-42a0-9e9c-e2b331a63fa9.png)
![image](https://user-images.githubusercontent.com/2420411/228078951-542d89cd-9214-4520-a39d-48cb6382b10e.png)

strategineer commented on src/help.cpp:
![image](https://user-images.githubusercontent.com/2420411/228079488-ec34c4dd-6d87-47d4-89ce-0e45d81bd94e.png)

strategineer commented on src/inventory_ui.cpp:
![image](https://user-images.githubusercontent.com/2420411/228079961-bfc0583b-b991-421f-a3b9-f2ad99123777.png)

strategineer commented on src/item_contents.cpp:
![image](https://user-images.githubusercontent.com/2420411/228080428-0caceafe-2e99-438c-ae29-7aafcc281f0b.png)

strategineer commented on src/iuse_software_kitten.cpp:
![image](https://user-images.githubusercontent.com/2420411/228080570-0fe03129-94c1-471f-ae5d-076b112c3d8c.png)
![image](https://user-images.githubusercontent.com/2420411/228080631-f347f069-5fbc-4e1b-95f1-e17857f0e90e.png)
![image](https://user-images.githubusercontent.com/2420411/228080646-1f2a21fc-c7ab-47b9-b2ab-919da2b96988.png)

strategineer commented on src/iuse_software_snake.cpp:
![image](https://user-images.githubusercontent.com/2420411/228080716-84ec3a0f-a02f-4c68-94f1-f0f62ece1642.png)

strategineer commented on src/iuse_software_lightson.cpp:
![image](https://user-images.githubusercontent.com/2420411/228080801-22dfc369-f784-450e-a5d7-0b014246d470.png)

strategineer commented on src/iuse_software_minesweeper.cpp:
![image](https://user-images.githubusercontent.com/2420411/228080865-759a4784-e927-4e00-83a8-40fbe03cddbd.png)

strategineer commented on src/iuse_software_sokoban.cpp:
![image](https://user-images.githubusercontent.com/2420411/228080938-c1d50150-9913-4bd9-82aa-c5d57a18a9c1.png)

strategineer commented on src/magic.cpp:
![image](https://user-images.githubusercontent.com/2420411/228082775-70d37ce7-2384-4692-a683-0c130b628eb7.png)
![image](https://user-images.githubusercontent.com/2420411/228082793-c042b15f-8117-45a9-83f6-e0324ee155ca.png)

strategineer commented on src/main_menu.cpp:
![image](https://user-images.githubusercontent.com/2420411/228082840-5fc44dc5-1e29-4e4d-b732-9e5ffc5c43c6.png)

strategineer commented on src/medical_ui.cpp:
![image](https://user-images.githubusercontent.com/2420411/228082972-bac96db2-06e9-44fb-8c4b-bf115a9ff527.png)

strategineer commented on src/messages.cpp:
![image](https://user-images.githubusercontent.com/2420411/228083025-46fd3acc-a2cd-4aef-8240-29a02cda591a.png)

strategineer commented on src/mutation_ui.cpp:
![image](https://user-images.githubusercontent.com/2420411/228083377-ddc38456-9336-4962-8fe0-a47978a9be8b.png)
![image](https://user-images.githubusercontent.com/2420411/228083415-8038cd66-428a-46c1-a60a-95967c2e339c.png)

strategineer commented on src/newcharacter.cpp:
![image](https://user-images.githubusercontent.com/2420411/228083532-6e9bcc7a-9a7d-430d-99e0-305256df1bec.png)
![image](https://user-images.githubusercontent.com/2420411/228083538-ffdd0db2-eaf7-4a35-b08c-806ced6e51de.png)
![image](https://user-images.githubusercontent.com/2420411/228083548-3de586ba-0db3-4f2c-b263-98a9b6f85ff9.png)
![image](https://user-images.githubusercontent.com/2420411/228083553-38b11e66-d595-4a74-95d8-feec69d1c1ee.png)
![image](https://user-images.githubusercontent.com/2420411/228083565-c94a58ee-0234-417e-b8fe-c5e05dcc6bcf.png)
![image](https://user-images.githubusercontent.com/2420411/228083576-2b53e4a6-e331-41f8-a2a0-839348171dea.png)
![image](https://user-images.githubusercontent.com/2420411/228083586-07fbaa08-f8ad-43b0-896d-62d7435e93d2.png)
![image](https://user-images.githubusercontent.com/2420411/228083595-6a094d7f-d424-4df4-b6c7-ecca4de69712.png)

strategineer commented on src/npctalk.cpp:
![image](https://user-images.githubusercontent.com/2420411/228083674-5e662e14-25e9-4aea-87b5-fd69d7dbcb50.png)

strategineer commented on src/options.cpp:
![image](https://user-images.githubusercontent.com/2420411/228083749-815d6b1d-b803-4b60-879b-aeeee6cea131.png)

strategineer commented on src/overmap_ui.cpp:
![image](https://user-images.githubusercontent.com/2420411/228083828-70c309a5-6bd9-4907-b265-b938f313da94.png)
![image](https://user-images.githubusercontent.com/2420411/228083865-907c866f-35fc-4a0f-83b5-3f5799bd13a7.png)

strategineer commented on src/panels.cpp:
![image](https://user-images.githubusercontent.com/2420411/228083982-9da2e35d-818e-48f5-9eb4-fa40a66c0366.png)

strategineer commented on src/player_display.cpp:
![image](https://user-images.githubusercontent.com/2420411/228084015-9d115c23-fc8d-416c-bf8e-c3546d1e8e38.png)

strategineer commented on src/popup.cpp:
![image](https://user-images.githubusercontent.com/2420411/228084081-95838e11-e38f-45ac-8323-c5ee37df93f9.png)

strategineer commented on src/proficiency_ui.cpp:
![image](https://user-images.githubusercontent.com/2420411/228084144-54dcbd9a-a4b8-426f-8eff-83390ee2ef61.png)

strategineer commented on src/ranged.cpp:
![image](https://user-images.githubusercontent.com/2420411/228084238-e8bc9918-86a3-4b6b-b98d-a0850e7f0eab.png)

strategineer replied:
![image](https://user-images.githubusercontent.com/2420411/228084310-849b6c5b-80ff-48c3-965f-f47a85f33966.png)

strategineer commented on src/safemode_ui.cpp:
![image](https://user-images.githubusercontent.com/2420411/228084375-1c235805-35da-41e1-b286-6e39df287d8a.png)

strategineer commented on src/smart_controller_ui.cpp:
![image](https://user-images.githubusercontent.com/2420411/228084610-609d1b11-153d-4a7e-a626-5a16e60902b2.png)

strategineer commented on src/trade_ui.cpp:
![image](https://user-images.githubusercontent.com/2420411/228084704-2a0d1cf6-b71c-4cbd-b019-b634da88c9e8.png)

strategineer commented on src/ui.cpp:
![image](https://user-images.githubusercontent.com/2420411/228085039-2326baea-7d2b-4690-8dd1-ad397e52f723.png)

strategineer committed:
Fix minor UI issues

strategineer commented on src/gamemode_defense.cpp:
![image](https://user-images.githubusercontent.com/2420411/228086048-b0276c31-0795-44bf-8de6-20d3a6e697ec.png)

strategineer committed:
Replace use of regex_replace based on feedback

strategineer committed:
Fix clang-tidy issues

Qrox commented:
Currently it looks a bit weird when using the inline format for descriptions containing spaces, especially when the descriptions are also separated by space. Perhaps disabling the inline format when a description contains space will make it look better?

Qrox commented on src/ui.cpp:
```diff
@@ -778,9 +779,13 @@ void uilist::show( ui_adaptor &ui )
             if( entries[ei].hotkey.has_value() && entries[ei].hotkey.value() != input_event() ) {
-                const nc_color hotkey_co = ei == selected ? hilight_color : hotkey_color;
-                mvwprintz( window, point( pad_left + 1, estart + si ), entries[ ei ].enabled ? hotkey_co : co,
-                           "%s", right_justify( entries[ei].hotkey.value().short_description(), 2 ) );
```
The hotkey description could be two letters long in the keycode mode (for example ctrl+a will be displayed as `^a`):

https://github.com/CleverRaven/Cataclysm-DDA/blob/6d311a79b1ba922a0bd86282b3f2290e47a81428/src/input.cpp#L159-L163

So you should keep `right_justify` and add one more to pad_left in line 797 below.

strategineer replied:
I think this did the trick https://github.com/CleverRaven/Cataclysm-DDA/pull/64478/commits/cdf95fbd01f4cab869bb3ccf1b10cd97e65acadd

Qrox replied:
I did my best to mark which changes are probably not needed, but the uilist code is a mess and I'm tired analyzing magic number now. Might want to fine-tune it later instead.

strategineer replied:
Thanks for being so meticulous. This https://github.com/CleverRaven/Cataclysm-DDA/pull/64478/commits/2fbc662ed6e32eee9b009acc12f23fd661beb2c0 should resolve the issue. I replaced my WIDTH_BUFFER variable back to 4 everywhere one-by-one, recompiled and tested a few uilists until I found which ones needed to actually be 6 now :) due to the added padding

Qrox commented on src/worldfactory.cpp:
```diff
@@ -1666,12 +1649,12 @@ int worldfactory::show_worldgen_basic( WORLD *world )
-                           ctxt.get_desc( "PICK_RANDOM_WORLDNAME", 1U ), ctxt.get_desc( "CONFIRM", 1U ),
-                           ctxt.get_desc( "HELP_KEYBINDINGS", 1U ) );
+            string_format( _( "Press %s to pick a random name for your world.\n"
+                              "Navigate options with directional keys "
```
Should print the actual hotkeys

Qrox commented on src/worldfactory.cpp:
```diff
@@ -1875,21 +1850,14 @@ void worldfactory::draw_modselection_borders( const catacurses::window &win,
-                    ctxtp.get_desc( "PREV_CATEGORY_TAB" ),
-                    ctxtp.get_desc( "NEXT_CATEGORY_TAB" ),
-                    ctxtp.get_desc( "HELP_KEYBINDINGS" )
+                    _( "%s | %s | %s | %s | %s" ),
```
Maybe remove `|` to be consistent with the style of other keybinding text?

strategineer committed:
Fix trying to inline keys into sentences (doesn't look good), add spacing to between hotkeys and response text in dialogues, fix help menu looking wonky

strategineer commented:
> Currently it looks a bit weird when using the inline format for descriptions containing spaces, especially when the descriptions are also separated by space. Perhaps it will look better if you disable the inline format when the description contains space and the hotkey is not the first character of the description?
> 
> The help menu also looks a bit messy with the inline format, might want to use separated format for it.
> 
> A space between the dialogue option hotkey and text might make it look better.

This should look better!

![image](https://user-images.githubusercontent.com/2420411/228285096-3e9d6474-3ddc-4dc9-9357-8eb212533a17.png)


![WindowsTerminal_qycuatLnpg](https://user-images.githubusercontent.com/2420411/228284850-cfc3c5eb-e51f-4164-9d46-a814a5ff36af.png)

strategineer committed:
Document scolorize function and refactor editmap code

strategineer committed:
If the first letter of a sentence is the same as the key we want to hint at, then use it anyway

strategineer commented on src/input.cpp:
```diff
@@ -1112,71 +1117,291 @@ std::string input_context::get_desc( const std::string &action_descriptor,
     for( const input_event &evt : events ) {
         if( is_event_type_enabled( evt.type ) && evt_filter( evt ) ) {
             na = false;
+            if( n_text == 0 || !should_inline_if_possible || !can_inline ) {
+                break;
+            }
```
I thought so too but I tried this a few days ago and realized that we're going through this loop for two reasons:
- to try and inline the key within the text
- to set the `na` variable to false if there's at least one key bound to the event

So moving this check outside the loop breaks the logic.

strategineer committed:
Make minor changes based on feedback

strategineer committed:
Implement and use get_hint_directions and get_hint_mouse to dispplay directional keys hints and mouse only hints respectively

strategineer commented on src/ranged.cpp:
```diff
@@ -3646,62 +3631,51 @@ void target_ui::draw_controls_list( int text_y )
     }
     if( is_mouse_enabled() ) {
         std::string move = _( "Mouse: LMB: Target, Wheel: Cycle," );
         std::string fire = _( "RMB: Fire" );
```
Bound to actions

I changed the code to this.
```
std::string move = string_format( _( "%s %s" ), ctxt.get_hint_mouse( "SELECT", _( "Target" ),
                                        move_hint_state ), ctxt.get_hint_mouse( "SCROLL_UP", _( "Cycle" ), move_hint_state ) );
        std::string fire = ctxt.get_hint_mouse( "FIRE", fire_hint_state );
```

and implemented the `get_hint_mouse` function in case we want to use them elsewhere eventually

strategineer commented on src/ranged.cpp:
```diff
@@ -3646,62 +3631,51 @@ void target_ui::draw_controls_list( int text_y )
+        lines.push_back( {2, ctxt.get_hint( "AIM", _( "Steady your aim (10 moves)" ), fire_hint_state )} );
+        lines.push_back( {2, ctxt.get_hint( "STOPAIM", _( "Stop aiming" ), fire_hint_state )} );
+        lines.push_back( {4, _( "Aim and fire with:" )} );
+        lines.push_back( {5, string_format( "  (%s)", enumerate_as_string( aim_and_fire, enumeration_conjunction::space ) )} );
```
huhmm I changed things around, it seems like it's working better now

![image](https://user-images.githubusercontent.com/2420411/228325169-89595548-3fec-47b7-b97c-a5b201a9361b.png)


![image](https://user-images.githubusercontent.com/2420411/228325124-dbaf9f4a-915f-41ba-ac63-69e578daef34.png)

strategineer committed:
More ranged menu changes

strategineer committed:
Fix trade ui

strategineer committed:
Fix up uilist display with proper padding / auto width calculation

strategineer commented on src/safemode_ui.cpp:
```diff
@@ -144,51 +144,49 @@ void safemode::show( const std::string &custom_name_in, bool is_safemode_in )
                 translate_marker( "<T>est" )
             }
         };
```
Good catch, it isn't. "<M>ove" is included, the action is called "SWAP_RULE_GLOBAL_CHAR"

strategineer commented on src/options.cpp:
```diff
@@ -1685,6 +1685,67 @@ void options_manager::add_options_interface()
+
+
+        add( "KEYBINDINGS_INLINE_UP_TO_X", page_id, to_translation( "Limit inlining of keybinding hints" ),
+             to_translation( "Inline keybindings in hints (like 'E(x)amine') if the length of the hint text is smaller than X characters. Disable this limit by setting this to 0." ),
```
deleting this option as your proposal to only inline the keybind when there's no spaces in the suffix/text part makes this option obsolete

strategineer committed:
Fix smart_controller_ui

strategineer committed:
Use get_hint_basic in replace_keybind_tag function

strategineer commented on src/game.cpp:
```diff
@@ -7333,24 +7328,16 @@ look_around_result game::look_around(
-                                                 ctxt.get_desc( "toggle_pixel_minimap" ),
-                                                 ctxt.get_action_name( "toggle_pixel_minimap" ) );
-                right_print( w_info, getmaxy( w_info ) - 1, 1, pixel_minimap_option ? c_light_green : c_green,
-                             pixel_minimap_text );
+                right_print( w_info, getmaxy( w_info ) - 1, 1, ctxt.get_hint( "toggle_pixel_minimap",
+                             pixel_minimap_option ? keybinding_hint_state::TOGGLED_ON : keybinding_hint_state::TOGGLED_OFF ) );
             }
```
I added an ifdef above to solve this problem
```
#if defined(TILES)
    ctxt.register_action( "toggle_pixel_minimap" );
#endif
```

when playing on curses, the keybind doesn't show up in the keybinds menu or this menu.

NB: This feature is also broken https://github.com/CleverRaven/Cataclysm-DDA/issues/58776 so that's something we'll tackle in another PR.

strategineer commented on src/armor_layers.cpp:
```diff
@@ -1175,49 +1172,32 @@ void outfit::sort_armor( Character &guy )
-                               ctxt.get_desc( "REMOVE_ARMOR" ) ),
-                "\n",
-                _( "Encumbrance explanation:\n" ),
+                ctxt.get_hint_pair( "UP", "DOWN", _( "to scroll the left pane (list of items)." ) ),
+                ctxt.get_hint_pair( "SCROLL_ITEM_INFO_UP", "SCROLL_ITEM_INFO_DOWN", _( "to scroll the middle pane (item information)." ) ),
+                ctxt.get_hint_pair( "PREV_TAB", "NEXT_TAB", _( "to scroll the right pane (items grouped by body part)." ) ),
+                ctxt.get_hint_pair( "LEFT", "RIGHT", _( "to limit the left pane to a particular body part." ) ),
+                ctxt.get_hint( "MOVE_ARMOR", _( "to select an item for reordering." ) ),
+                ctxt.get_hint( "ASSIGN_INVLETS", _( "to assign special inventory letters to clothing." ) ),
+                ctxt.get_hint( "CHANGE_SIDE", _( "to change the side on which item is worn." ) ),
+                ctxt.get_hint( "TOGGLE_CLOTH", _( "to toggle item visibility on character sprite." ) ),
+                ctxt.get_hint( "SORT_ARMOR", _( "to sort worn items into natural layer order." ) ),
+                ctxt.get_hint( "EQUIP_ARMOR", _( "to equip a new item." ) ),
+                ctxt.get_hint( "EQUIP_ARMOR_HERE", _( "to equip a new item at the currently selected position." ) ),
+                ctxt.get_hint( "REMOVE_ARMOR", _( "to remove selected item." ) ),
```
I resolved this by implementing your proposal to not inline keys into sentences.

strategineer committed:
Fix minor issues

strategineer committed:
Fix main menu items spacing

strategineer committed:
Fix

strategineer commented on src/main_menu.cpp:
```diff
@@ -348,7 +350,7 @@ void main_menu::print_menu( const catacurses::window &w_open, int iSel, const po
     int menu_length = 0;
     for( size_t i = 0; i < vMenuItems.size(); ++i ) {
-        menu_length += utf8_width_notags( vMenuItems[i].c_str() ) + 2;
+        menu_length += utf8_width_notags( vMenuItems[i].c_str() ) + 3;
```
I think I was able to resolve this but i'll need to take another look.

strategineer commented on src/overmap_ui.cpp:
```diff
@@ -1158,27 +1158,33 @@ static void draw_om_sidebar(
-    mvwprintz( wbar, point( 1, 13 ), c_magenta, _( "Press W to preview route." ) );
-    mvwprintz( wbar, point( 1, 14 ), c_magenta, _( "Press again to confirm." ) );
+    mvwprintz( wbar, point( 1, 12 ), input_context::get_hint_color(),
+               _( "Use movement keys to pan." ) );
```
![image](https://user-images.githubusercontent.com/2420411/228407963-f64a50f0-2efc-4c51-a2a9-9a7b93b67040.png)

strategineer committed:
Use actual directional keys in hint in overmap_ui

strategineer committed:
Reuse replace_substring instead of re-implementing a similar function

strategineer committed:
Clean up padding with non breaking space code

strategineer committed:
Fix how we deal with padding the suffix and don't inline keybinds ith spaces in them (even if the key is the first letter of the sentence)

Qrox commented on src/ui.cpp:
```diff
@@ -538,20 +543,21 @@ void uilist::calc_data()
+            if( w_auto && w_width < txtwidth + pad + WIDTH_BUFFER + clen ) {
                 // TODO: or +5 if header
-                w_width = txtwidth + pad + 4 + clen;
+                w_width = txtwidth + pad + WIDTH_BUFFER + clen;
             }
         }
         if( desc_enabled ) {
-            const int min_desc_width = std::min( max_desc_width, std::max( w_width, descwidth_final ) - 4 );
+            const int min_desc_width = std::min( max_desc_width, std::max( w_width,
+                                                 descwidth_final ) - WIDTH_BUFFER );
             int descwidth = find_minimum_fold_width( footer_text.empty() ? entries[i].desc : footer_text,
                             desc_lines, min_desc_width, max_desc_width );
-            descwidth += 4; // 2x border + 2x ' ' pad
+            descwidth += WIDTH_BUFFER; // 2x border + 4x ' ' pad
```
Might not be needed because description does not start with a hotkey hint.

strategineer replied:
resolved by https://github.com/CleverRaven/Cataclysm-DDA/pull/64478/commits/2fbc662ed6e32eee9b009acc12f23fd661beb2c0

Qrox commented on src/ui.cpp:
```diff
@@ -607,15 +613,15 @@ void uilist::calc_data()
-            if( realtextwidth + 4 > w_width ) {
-                w_width = realtextwidth + 4;
+            if( realtextwidth + WIDTH_BUFFER > w_width ) {
+                w_width = realtextwidth + WIDTH_BUFFER;
```
I think `text` is printed before all uilsit entries and does not start with a hotkey hint, so changing 4 to `WIDTH_BUFFER` is not needed.

Qrox commented on src/ui.cpp:
```diff
@@ -627,9 +633,8 @@ void uilist::calc_data()
-            // -2 for borders, -2 for padding
             desc_lines = std::max<int>( desc_lines, foldstring( footer_text.empty() ? ent.desc : footer_text,
-                                        w_width - 4 ).size() );
+                                        w_width - WIDTH_BUFFER ).size() );
```
Not needed for description.

strategineer replied:
resolved by https://github.com/CleverRaven/Cataclysm-DDA/pull/64478/commits/2fbc662ed6e32eee9b009acc12f23fd661beb2c0

Qrox commented on src/ui.cpp:
```diff
@@ -692,7 +697,7 @@ void uilist::reposition( ui_adaptor &ui )
     setup();
     if( filter_popup ) {
-        filter_popup->window( window, point( 4, w_height - 1 ), w_width - 4 );
+        filter_popup->window( window, point( WIDTH_BUFFER, w_height - 1 ), w_width - WIDTH_BUFFER );
```
Not needed for filter.

Qrox commented on src/ui.cpp:
```diff
@@ -742,7 +747,7 @@ void uilist::show( ui_adaptor &ui )
     if( !textformatted.empty() ) {
         for( int i = 0; i < text_lines; i++ ) {
-            trim_and_print( window, point( 2, 1 + i ), getmaxx( window ) - 4,
+            trim_and_print( window, point( 2, 1 + i ), getmaxx( window ) - WIDTH_BUFFER,
```
`text` starts one space after the left border and end one space before the right border, so no need to change this.

strategineer replied:
resolved by https://github.com/CleverRaven/Cataclysm-DDA/pull/64478/commits/2fbc662ed6e32eee9b009acc12f23fd661beb2c0

Qrox commented on src/ui.cpp:
```diff
@@ -836,7 +843,7 @@ void uilist::show( ui_adaptor &ui )
 
         if( static_cast<size_t>( selected ) < entries.size() ) {
-            fold_and_print( window, point( 2, w_height - desc_lines - 1 ), w_width - 4, text_color,
+            fold_and_print( window, point( 2, w_height - desc_lines - 1 ), w_width - WIDTH_BUFFER, text_color,
```
Not needed for description.

strategineer committed:
Remove obsolete comment

strategineer committed:
Clean up help menu shortcut/section display

strategineer commented on src/npctalk.cpp:
```diff
@@ -1818,7 +1818,7 @@ talk_data talk_response::create_option_line( const dialogue &d, const input_even
     talk_data results;
     results.color = color;
-    results.hotkey_desc = right_justify( hotkey.short_description(), 2 );
+    results.hotkey_desc = hotkey.short_description();
```
How would you get this to show a key with a modifier? It seems like keys show up alphabetically from a to z?

strategineer commented on src/overmap_ui.cpp:
```diff
@@ -1158,8 +1158,7 @@ static void draw_om_sidebar(
 
-    mvwprintz( wbar, point( 1, 12 ), input_context::get_hint_color(),
-               _( "Use movement keys to pan." ) );
+    print_colored_text( wbar, point( 1, 12 ), inp_ctxt->get_hint_directions( _( "Pan view" ) ) );
```
Those are keys I'm using for moving around in cardinal directions (I've got a weird keyboard layout).

With default settings it shows the following:
![image](https://user-images.githubusercontent.com/2420411/228556277-4d86cc02-54cd-4327-97ae-4db0f4ea5e56.png)

Which seems reasonable although I suspect that most players don't play with vi-keys and use the keypad instead (or maybe even arrow keys and avoid moving diagonally). It's hard to know what players are actually using to move around and thus hard to know what we should display here.

I'm starting to think that because movement like this is the default "action" players are doing, we can avoid displaying any sort of keybinds related to this because if the player has gotten this far, they know how to move around their view/character (however they like to do so) and we probably don't need to remind them.

What do you think?

Qrox replied:
That's actually a good point.

Qrox replied:
I don't have a strong preference for either so I'll leave that up to you.

strategineer committed:
Add todo

strategineer commented on src/ui.cpp:
```diff
@@ -783,20 +788,22 @@ void uilist::show( ui_adaptor &ui )
                 entries[ei].drawn_rect =
-                    inclusive_rectangle<point>( p + point( -3, 0 ), p + point( -4 + pad_size, 0 ) );
+                    inclusive_rectangle<point>( p + point( -WIDTH_BUFFER, 0 ), p + point( -WIDTH_BUFFER + pad_size,
+                                                0 ) );
```
![WindowsTerminal_N7YfYSUBv7](https://user-images.githubusercontent.com/2420411/228558575-59239abb-731a-4b7f-8bd8-724dfb4777f3.gif)

It seems good!

strategineer commented on src/ui.cpp:
```diff
@@ -850,7 +857,7 @@ void uilist::show( ui_adaptor &ui )
         if( !filter.empty() ) {
             mvwprintz( window, point( 2, w_height - 1 ), border_color, "< %s >", filter );
-            mvwprintz( window, point( 4, w_height - 1 ), text_color, filter );
+            mvwprintz( window, point( WIDTH_BUFFER, w_height - 1 ), text_color, filter );
```
Yes I broke this definitely :).

strategineer committed:
Fix padding

strategineer commented on src/ui.cpp:
```diff
@@ -495,14 +500,14 @@ void uilist::calc_data()
         w_width = w_width_setup.fun();
     }
-    const int max_desc_width = w_auto ? TERMX - 4 : w_width - 4;
+    const int max_desc_width = w_auto ? TERMX - WIDTH_BUFFER : w_width - WIDTH_BUFFER;
```
resolved by https://github.com/CleverRaven/Cataclysm-DDA/pull/64478/commits/2fbc662ed6e32eee9b009acc12f23fd661beb2c0

strategineer committed:
Pad uilist hotkey properly

strategineer committed:
Display main menu world sub menu consistently

strategineer committed:
Refactor get_hint_pair and get_hint_quad functions into get_hints, and avoid hacky renaming of mouse buttons

strategineer committed:
Remove regex include

strategineer commented on src/input.cpp:
```diff
@@ -10,6 +10,7 @@
 #include <memory>
 #include <new>
 #include <optional>
+#include <regex>
```
I thought I removed this... Sorry about that.

strategineer commented on src/input.cpp:
```diff
@@ -1105,7 +1110,20 @@ std::string input_context::get_desc( const std::string &action_descriptor,
     for( size_t i = 0; i < inputs_to_show.size(); ++i ) {
-        rval += inputs_to_show[i].long_description();
+        const input_event &event = inputs_to_show[i];
+        std::string description  = event.long_description();
```
I reverted my hacky change here. I'll leave the strings as-is for now unless someone complains about "MOUSE_LEFT" looks weird in the UI.

strategineer committed:
Make query_yn options consistently colored

strategineer commented on src/popup.cpp:
```diff
@@ -103,8 +103,13 @@ std::vector<std::vector<std::string>> query_popup::fold_query(
+        //   problem with the default settings because the key is separated
+        //   by parens but still...
+        const std::string &desc = ctxt.get_hint( opt.action, keybinding_hint_state::NONE_AT_ALL,
+                                  opt.filter );
```
it's a little hacky but my solution here works https://github.com/CleverRaven/Cataclysm-DDA/pull/64478/commits/cdadc091f83a81f5d886ec7e31d3d06b68b4d80b

strategineer replied:
I refactored the code so it's a little less hacky now.

I couldn't quite get the other solution to work cleanly (although I understand it's possible). Do you think this can be good enough for now and I can come back to this in another PR? Or is this unmergeable in this state you think?

Qrox replied:
If you add a comment to the replacement function to warn that it could break in some edge cases I think it's good enough for now.

Qrox commented on src/main_menu.cpp:
```diff
@@ -350,14 +350,13 @@ void main_menu::print_menu( const catacurses::window &w_open, int iSel, const po
-        }
+        const std::string shortcut = remove_color_tags( shortcut_text( vMenuItems[i] ) );
+        const int shortcut_width = utf8_width_notags( shortcut.c_str() ) + 1;
+        menu_length += shortcut_width;
```
Oh, nevermind then, not sure how I missed it.

Qrox commented on src/main_menu.cpp:
```diff
@@ -348,14 +351,13 @@ void main_menu::print_menu( const catacurses::window &w_open, int iSel, const po
-            menu_length += utf8_width( vMenuHotkeys[i][0] );
-        }
+        const std::string shortcut = remove_color_tags( shortcut_text( vMenuItems[i] ) );
+        const int shortcut_width = utf8_width_notags( shortcut.c_str() ) + 1;
```
The following might fix the centering problem:
```suggestion
        const int shortcut_width = utf8_width_notags( shortcut.c_str() );
```

strategineer replied:
that combined with your other fix did the trick

![image](https://user-images.githubusercontent.com/2420411/228954437-7e5c4020-bf17-4451-bbd1-e81e79a7600b.png)


it works! ;)

Qrox commented on src/main_menu.cpp:
```diff
@@ -348,14 +351,13 @@ void main_menu::print_menu( const catacurses::window &w_open, int iSel, const po
     const int free_space = std::max( 0, window_width - menu_length - offset.x );
-    const int spacing = free_space / ( static_cast<int>( vMenuItems.size() ) + 1 );
-    const int width_of_spacing = spacing * ( vMenuItems.size() + 1 );
+    const int spacing = free_space / static_cast<int>( vMenuItems.size() );
+    const int width_of_spacing = spacing * vMenuItems.size();
```
```suggestion
    const int spacing = std::max( 1, free_space / static_cast<int>( vMenuItems.size() + 1 ) );
    const int width_of_spacing = spacing * ( vMenuItems.size() + 1 );
```

Qrox commented on src/main_menu.cpp:
```diff
@@ -119,8 +119,8 @@ std::vector<int> main_menu::print_menu_items( const catacurses::window &w_in,
         ret.push_back( utf8_width_notags( text.c_str() ) );
 
-        std::string temp = shortcut_text( iSel == i ? hilite( c_yellow ) : c_yellow, vItems[i] );
-        text += string_format( "[%s]", colorize( temp, iSel == i ? hilite( c_white ) : c_white ) );
```
The following call at line 131 should be updated to use the default separator (remove `']'`) because the items are now not separated by `]` anymore.

Qrox reviewed (commented):
Some of these are responses to previous comments but somehow github decides to show them as standalone comments while burying the context in the comment history.

strategineer committed:
Fix main menu menu item centering

strategineer commented on src/ranged.cpp:
```diff
@@ -3534,63 +3519,48 @@ void target_ui::draw_controls_list( int text_y )
-        std::string move = _( "Mouse: LMB: Target, Wheel: Cycle," );
-        std::string fire = _( "RMB: Fire" );
-        lines.push_back( {7, colored( col_move, move ) + " " + colored( col_fire, fire )} );
+        std::string move = string_format( _( "%s %s" ), ctxt.get_hint_mouse( "SELECT", _( "Target" ),
+                                          move_hint_state ), ctxt.get_hint_mouse( "SCROLL_UP", _( "Cycle" ), move_hint_state ) );
+        std::string fire = ctxt.get_hint_mouse( "FIRE", fire_hint_state );
+        lines.push_back( {8, string_format( "%s %s", move, fire )} );
```
These hints were displaying mouse inputs specifically. I've retained that by limiting the hints in this way because the keys to move/fire with the keyboard are displayed further down. What would like me to do here?

Qrox replied:
Previously I implemented selecting with the mouse cursor and a keyboard key by binding a non-mouse key to the "SELECT" action, which is different from using the keyboard hotkey to fire at the currently selected target. Could you try rebinding a keyboard hotkey to the "SELECT" action and see the hints make sense in that case?

strategineer committed:
Call foldstring with default separator because ']' isn't used toto separate menu items anymore

strategineer commented on src/ranged.cpp:
```diff
@@ -3534,63 +3519,48 @@ void target_ui::draw_controls_list( int text_y )
-    }
+    std::string movement_text = shifting_view ? _( "Shift view" ) : _( "Move cursor" );
+    movement_text = colorize( movement_text, input_context::get_hint_color( move_hint_state ) );
+    lines.push_back( {9, string_format( "%s %s", ctxt.get_hint_directions( move_hint_state ), movement_text )} );
```
Definitely!

strategineer committed:
Refactor

strategineer committed:
Fix clang-tidy issue

strategineer commented on src/output.cpp:
```diff
@@ -2608,20 +2608,18 @@ void print_global_and_character_mode_headers( const catacurses::window &w, const
+        std::string hint_text = string_format( "%s%s%s", prestring,
+                                               remove_shortcut_from_text ? "" : shortcut,
                                                poststring );
         return input_context::get_hint_basic( shortcut, hint_text, state );
```
> Well, I thought removing the colon would make it replace the first letter with the hotkey even if the string contains spaces but I forgot you reverted that change.

It did do that. The menu looked OK but it used the inlined separators instead of the separate separators ("(a): Blah" instead of "[a]: Blah"). This wasn't consistent so I added this extra `remove_shortcut_from_text ` to the `shortcut_text` function as a way of dealing with this edge case. The code works as expected in every other context the `shortcut_text` function is used (I'm pretty sure) because the `remove_shortcut_from_text` param defaults to false.

For example, this is what the french menu looks like:
![image](https://user-images.githubusercontent.com/2420411/228965118-4394658c-d1ee-4e51-b27f-0b218283ca88.png)

Just to clarify, are there other reasons to implement the solution you've described as opposed to what has already been implemented?

Qrox replied:
Can you test the Japanese menu and let me know if it prints the correct strings? They translated `<W|w> world` as `<W|w> 世界` and I'm concerned it may become `[W] W 世界` with the current code.

strategineer replied:
Good catch, ya this is broken.

![image](https://user-images.githubusercontent.com/2420411/229115655-3c808ed7-4bd9-41b6-bb7f-f5919479a2d7.png)

This change should resolve the issue but I'll need to test this more. https://github.com/CleverRaven/Cataclysm-DDA/pull/64478/commits/394e371de9f1ff2c575251637b6d0b9aef1bae8f

There's a lot of edge cases ;)

Qrox replied:
Please test the Chinese (simplified and traditional) menu with inline format disabled in the settings. They removed the space from the translation (so `<W|w> World` becomes `<W|w>世界` instead of `<W|w> 世界`) and I think the newly added checks will not work in that case.

Although my point is that it's hard to add checks for data that are not (fully) controled by the project itself so I think it may still be more approriate to only replace the hotkey tags with the formatted hotkeys in place and disregard the user's inline setting, at least for now until we fully migrate the main menu hotkey system to be more in line with the rest of the game.

strategineer committed:
Make vehicle scrolling keybindings consistent. Fixes #59698

strategineer committed:
Improve shortcut_text function to work better with more kinds of inputs
